library/repair_docker_data_volumes

```diff
@@ -0,0 +1,338 @@
+  - The secondary case is where the container did not own the volumes, but was pulling them
+    in from another container using "--volumes-from". In this case, the "Volumes" and
+    "VolumesRW" keys appear to be removed during the update process, as well as the
+    "AppliedVolumesFrom" key. In order to address this case, we rely on configuration data
+    collected before the package update (by the collect_container_configs module).
+requirements:
+  - this module should be run after the docker package upgrade
+  - if changes happen, the docker daemon should be started again.
+  - i.e. run this with "notify: restart docker"
+options:
+  fix_vfs_links:
+    description:
+      - should we fix the VFS links (issue 1)?
+    requried: true
+  fix_container_mountpoints:
+    description:
+      - should we fix the container mountpoints (issue 2)?
+    required: true
+  old_container_configs:
+    description:
+      - config data from containers, before the package update
+'''
+
+CONTAINER_ROOT_DIR = '/var/lib/docker/containers'
+CONFIG_V2_FILENAME = 'config.v2.json'
+CONFIG_V3_FILENAME = 'config.v3.json'
+BACKUP_FILENAME = 'config.v2.json.pre_volume_fix'
+VFS_BASE_DIR = '/var/lib/docker/vfs/dir'
+VOLUMES_ROOT_DIR = '/var/lib/docker/volumes'
+LOCAL_DRIVER = 'local'
+
+VFS_MOUNT_REGEX = re.compile(VFS_BASE_DIR + "/([a-f0-9]+)$")
+VOLUME_MOUNT_REGEX = re.compile(VOLUMES_ROOT_DIR + "/([a-f0-9]+)/_data")
+
+
+class Volume(object):
+    def __init__(self, driver, name):
+        self.driver = driver
+        self.name = name
+
+    def inspect(self, module):
+        try:
+            inspect_output = subprocess.check_output(
+                ['docker', 'volume', 'inspect', self.name],
+                stderr=subprocess.STDOUT)
+            return json.loads(inspect_output)[0]
+        except subprocess.CalledProcessError, c:
+            module.fail_json(
+                msg="failure while inspecting volume {}".format(self.name),
+                exception=str(c))
+
+    @property
+    def is_local(self):
+        return self.driver == LOCAL_DRIVER
+
+    @property
+    def vfs_path(self):
+        return os.path.join(VFS_BASE_DIR, self.name)
+
+
+def _volumes(module):
+    try:
+        volumes_output = subprocess.check_output(
+            ['docker', 'volume', 'ls'],
+            stderr=subprocess.STDOUT)
+
+        return [
+            Volume(driver=line.split()[0], name=line.split()[1])
+            for line in volumes_output.split('\n')[1:] if line != '']
+
+    except subprocess.CalledProcessError, c:
+        module.fail_json(
+            msg="failure while listing volumes",
+            exception=str(c))
+
+
+class CreateSymlinkOperation(object):
+    def __init__(self, source_path, link_path):
+        self.source_path = source_path
+        self.link_path = link_path
+
+    def run(self, module):
+        subprocess.check_call(['ln', '-sfn', self.source_path, self.link_path])
```

Comment:
Good improvement - thought I think no need for unlink, we already check if the link is present, and if there is something there I think failing is better.